src/controller/java/src/matter/controller/MatterControllerImpl.kt

```diff
@@ -0,0 +1,661 @@
+                  }
+
+                  for (eventList in cluster.value.events) {
+                    for (event in eventList.value) {
+                      val timestamp: Timestamp =
+                        when (event.timestampType) {
+                          chip.devicecontroller.model.EventState.MILLIS_SINCE_BOOT ->
+                            Timestamp.MillisSinceBoot(event.timestampValue)
+                          chip.devicecontroller.model.EventState.MILLIS_SINCE_EPOCH ->
+                            Timestamp.MillisSinceEpoch(event.timestampValue)
+                          else -> {
+                            logger.log(Level.SEVERE, "Unsupported event timestamp type - ignoring")
+                            break
+                          }
+                        }
+
+                      val eventPath =
+                        EventPath(
+                          endpointId = endpoint.key.toUShort(),
+                          clusterId = cluster.key.toUInt(),
+                          eventId = eventList.key.toUInt()
+                        )
+
+                      val readData =
+                        ReadData.Event(
+                          path = eventPath,
+                          eventNumber = event.eventNumber.toULong(),
+                          priorityLevel = event.priorityLevel.toUByte(),
+                          timeStamp = timestamp,
+                          data = event.tlvValue
+                        )
+                      successes.add(readData)
+                    }
+                  }
+                }
+              }
+
+              trySendBlocking(SubscriptionState.NodeStateUpdate(ReadResponse(successes, failures)))
+                .onFailure { ex ->
+                  logger.log(Level.SEVERE, "Error sending NodeStateUpdate to subscriber: %s", ex)
+                }
+            }
+
+            override fun onError(
+              attributePath: ChipAttributePath?,
+              eventPath: ChipEventPath?,
+              ex: Exception
+            ) {
+              attributePath?.let {
+                logger.log(Level.INFO, "Report error for attributePath:%s", it.toString())
+                val tmpAttributePath: AttributePath = attributePath.wrap()
+                val attributeFailure = ReadFailure.Attribute(path = tmpAttributePath, error = ex)
+                failures.add(attributeFailure)
+              }
+              eventPath?.let {
+                logger.log(Level.INFO, "Report error for eventPath:%s", it.toString())
+                val tmpEventPath: EventPath = eventPath.wrap()
+                val eventFailure = ReadFailure.Event(path = tmpEventPath, error = ex)
+                failures.add(eventFailure)
+              }
+
+              // The underlying subscription is terminated if both attributePath & eventPath are
+              // null
+              if (attributePath == null && eventPath == null) {
+                trySendBlocking(
```

Comment:
maybe add an extra log here since this seems like an unusual state and I am unsure how callers generally handle errors. May as well single out this kind of state.